pkg/util/assertion/assertion.go

```diff
@@ -0,0 +1,219 @@
+//
+// Typical usage:
+//
+//	if foo != bar {
+//	    return assertion.Failed(ctx, "oh no: %v != %v", foo, bar)
+//	}
+//
+// (the API is not e.g. Assert(foo == bar, "oh no: %v != %v", foo, bar) since
+// this incurs unconditional runtime overhead for the format interface boxing)
+//
+// In non-release builds (including roachprod and roachtest clusters), assertion
+// failures kill the process with a stack trace to ensure failures are noticed.
+//
+// In release builds, assertion failures are not fatal, since this is often
+// excessive and causes unnecessary disruption and outages. Instead, an error is
+// logged (with stack trace), reported to Sentry, and returned to the caller.
+// The caller must explicitly handle the error (enforced by the errcheck
+// linter), and should typically propagate it, but can also choose to either
+// ignore or recover from the error (when safe), or unconditionally kill the
+// process if necessary even in release builds.
+//
+// This can also be controlled explicitly via COCKROACH_FATAL_ASSERTIONS.
+//
+// Assertions are used to check and enforce internal invariants, not for general
+// error handling or fatal errors. As a rule of thumb, only use an assertion for
+// something we expect will never ever fail. Some examples:
+//
+//   - Good assertion: a Raft range's start key must be before its end key. This
+//     must always be true, so this should be an assertion.
+//
+//   - Bad assertion: writing to a file must succeed. We expect this to fail
+//     sometimes, e.g. with faulty disks, so this should be a (possibly fatal)
+//     error, not an assertion.
+//
+//   - Worse assertion: users must enter a valid SQL query. We expect users to
+//     get this wrong all the time, so this should return an error. In general,
+//     assertions should never be used to validate external inputs.
+//
+// Sometimes, it may be appropriate to panic or fatal on assertion failures even
+// in release builds. Panics are often used to propagate errors in SQL code.
+// Fatals may be necessary when it is unsafe to keep the node running. The
+// helpers Panic() and Fatal() can be used for these cases, but only sparingly.
+//
+// Some assertions may be too expensive to always check. These can be gated on
+// ExpensiveEnabled, which is set to true in special test builds (currently race
+// builds), and will otherwise compile the assertion out entirely.
+//
+// USAGE EXAMPLES
+// ==============
+//
+// Example: ignoring an assertion failure. Here, RPC batch processing should
+// never return both a Go error and a response error. In release builds, we're
+// ok with just ignoring this and using the Go error.
+//
+//	func (n *Node) Batch(ctx context.Context, req *kvpb.BatchRequest) *kvpb.BatchResponse {
+//	  br, err := n.batchInternal(ctx, req)
+//	  if err != nil {
+//	    if br.Error != nil {
+//	      _ = assertion.Failed(ctx, "returned both br.Err=%v and err=%v", br.Err, err)
+//	    }
```

Comment:
No, but this will also log the assertion failure to the log and notify Sentry -- even if we don't kill the process, we still want to know that it happened.